#include "chatwin.h"
#include "ui_chatwin.h"
#include <QStyleOption>
#include <QPainterPath>
#include <QKeyEvent>
#include <QObject>
#include <QShortcut>
#include <QScrollBar>
#include <QDesktopServices>
#include <QDebug>
#include <QDateTime>
#include <QStringListModel>
#include <QSound>

chatwin::chatwin(QWidget *parent, QString Name) :
    QWidget(parent),
    ui(new Ui::chatwin),
    name(Name)
{
    //ÂàùÂßãÂåñ
    this->initUI();
    //ÂÖ®ÈÉ®connectÂáΩÊï∞
    this->allConnect();
    //styleÊñá‰ª∂
    QFile file("://res/styles/style_chat.css");
    file.open(QFile::ReadOnly);
    if(!file.isOpen()){
        QMessageBox::warning(this,"","Ê†∑ÂºèÂä†ËΩΩÂ§±Ë¥•ÔºåÊñá‰ª∂‰∏çÂ≠òÂú®");
    }
    //ËÆæÁΩÆstyle
    this->setStyleSheet(file.readAll());
}

void chatwin::initUI()
{
    ui->setupUi(this);
    this->setWindowTitle("NewChat");
    //Á™óÂè£ÂàùÂßãÂåñ
    this->setWindowIcon(QIcon(":/res/icons/icon.png"));
    //ÈÄèÊòéËÉåÊôØ
    setAttribute(Qt::WA_TranslucentBackground, true);
    setWindowFlags(Qt::FramelessWindowHint);
    //Âú®Á∫øÁä∂ÊÄÅÂõæÁâáÂ§ßÂ∞èÁº©Êîæ
    ui->online->setScaledContents(true);

    //ÂàùÂßãÂåñÂÆ¢Êà∑Á´ØËÅäÂ§©ÊòµÁß∞
    class thread proc_client_init(&chatwin::Client_init,this, this->name);
    proc_client_init.detach();

    //ÊúÄÂ∞èÂåñÂõæÊ†á
    QPixmap picMin(":/res/icons/minblue.png");
    picMin.scaled(30,40);
    ui->min_btn->setIcon(picMin);
    //ÂÖ≥Èó≠ÂõæÊ†á
    QPixmap picClose(":/res/icons/closeblue.png");
    picClose.scaled(30,40);
    ui->close_btn->setIcon(picClose);
    ui->select_line->setPlaceholderText("üîçÔ∏é ÊêúÁ¥¢");
    // Ê†πÊçÆÈúÄË¶ÅÈöêËóèÊ∞¥Âπ≥ÊªöÂä®Êù°
    ui->listWidget->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    // ËÅäÂ§©Ê°ÜÊªöÂä®Êù°ÂØπË±°
    vScrollbar = new QScrollBar(Qt::Vertical, ui->listWidget);
    vScrollbar->setObjectName("vScrollbar");
    // Â∞ÜÂûÇÁõ¥ÊªöÂä®Êù°ËÆæÁΩÆÂà∞QTableWidget‰∏≠
    ui->listWidget->setVerticalScrollBar(vScrollbar);

}

chatwin::~chatwin()
{
    delete ui;
    closesocket(fd);
    close_Socket();
}

void chatwin::get_name(QString Name)
{
    this->name=Name;
}
/*----------------ÂÆ¢Êà∑Á´ØÂàùÂßãÂåñ-------------------*/
void chatwin::Client_init(QString Name)
{
    init_Socket();
    char recvBuf[1024];
    fd = create_clientSocket("127.0.0.1");
    if(fd == INVALID_SOCKET){
        offline();
    }
    while(fd == INVALID_SOCKET){
        fd = create_clientSocket("127.0.0.1");
    }
    online();
    const char *Buf = new char[1024];
    Buf = Name.toUtf8().constData();
    strncpy(recvBuf, Buf,sizeof(recvBuf));
    recvBuf[sizeof(recvBuf) - 1] = '\0';
    int ret = send(fd, recvBuf, strlen(recvBuf), 0);
    if (ret <= 0) {
        err("send");

        //closesocket(fd);
    }
        class thread proc_client(&chatwin::child_fun,this, fd);
        proc_client.detach();
}
/*---------------ÂÆ¢Êà∑Á´ØÂèëÈÄÅÊ∂àÊÅØ-------------------*/
void chatwin::client_sent(QString buf)
{
    const char* Buf = new char[1024];
    Buf = buf.toUtf8().constData();
    char recvBuf[1024];
    strncpy(recvBuf, Buf,sizeof(recvBuf));
    recvBuf[sizeof(recvBuf) - 1] = '\0';
    int ret = send(fd, recvBuf, strlen(recvBuf), 0);
    if (ret <= 0) {
        err("send");
    }
}

void chatwin::child_fun(SOCKET fd)
{
    int ret;
    char recvBuf[1024];

    while (1) {
        memset(recvBuf, 0, sizeof(recvBuf));
        ret = recv(fd, recvBuf, sizeof(recvBuf), 0);

        if (ret <= 0) {
            err("recv");
//            class thread proc_client_init(&chatwin::Client_init,this, this->name);
//            proc_client_init.detach();
//            break;
            continue;
        }
        else{
            QString strl = QString::fromUtf8(recvBuf);
            qDebug()<<str<<endl;
            emit resultReady_She(strl);
        }
    }
    closesocket(fd);
}

/*------------------ÁªòÂà∂Á™óÂè£ÂúÜËßí-------------------*/
void chatwin::paintEvent(QPaintEvent *event)
{
    QPainter painter(this);
        painter.setRenderHint(QPainter::Antialiasing);    // ÂèçÈîØÈΩø;
        painter.setBrush(QBrush(Qt::white));
        painter.setPen(Qt::transparent);
        QRect rect = this->rect();
        rect.setWidth(rect.width() - 1);
        rect.setHeight(rect.height() - 1);
        painter.drawRoundedRect(rect, 15, 15);
        //‰πüÂèØÁî®QPainterPath ÁªòÂà∂‰ª£Êõø painter.drawRoundedRect(rect, 15, 15);
        /*{
            QPainterPath painterPath;
            painterPath.addRoundedRect(rect, 15, 15);

        }*/
        QWidget::paintEvent(event);
        QStyleOption opt;
            opt.initFrom(this);
            QPainter p(this);
            style()->drawPrimitive(QStyle::PE_Widget, &opt, &p, this);
            QWidget::paintEvent(event);
            painter.end();


}
/*------------------ËæìÂÖ•Ê°ÜÂèëÈÄÅ-------------------*/
void chatwin::allConnect()
{
    //Ëá™Áî±ÊãñÂä®
    connect(this, &chatwin::dragSignal, this, &chatwin::onDrag);

    // ÂàõÂª∫ÂõûËΩ¶ÈîÆÁöÑÂø´Êç∑ÈîÆÔºåÂ∞ÜÂÖ∂ËøûÊé•Âà∞ÊåâÈíÆÁöÑÁÇπÂáªÊßΩ
    connect(ui->plainTextEdit, &QPlainTextEdit::textChanged, [this]() {
        QString text = ui->plainTextEdit->toPlainText();
        // Ê£ÄÊü•ÊòØÂê¶Êåâ‰∏ã‰∫Ü Shift ÈîÆ
        Qt::KeyboardModifiers modifiers = QGuiApplication::keyboardModifiers();
        bool shiftPressed = (modifiers & Qt::ShiftModifier);
        if (text.endsWith('\n')&&!shiftPressed) {
            // ÁßªÈô§ÊúÄÂêé‰∏Ä‰∏™Â≠óÁ¨¶ÔºàÂç≥Êç¢Ë°åÁ¨¶Ôºâ
            text.chop(1);
            ui->plainTextEdit->setPlainText(text);
            // ÂèëÈÄÅ
            ui->enter_btn->click();
        }
    });
    //Êé•Êî∂‰∏éÂèëÈÄÅÊ∂àÊÅØÊòæÁ§∫
    connect(this,&chatwin::resultReady_She,this,&chatwin::recvMsg);
    connect(this,&chatwin::resultReady_Me,this,&chatwin::sendMsg);

}
/*---------------ÊúçÂä°Âô®Âú®Á∫øÁä∂ÊÄÅÊòæÁ§∫----------------*/
void chatwin::online()
{
    //Âú®Á∫øÁä∂ÊÄÅ
    QPixmap pixmap("://res/icons/online.png");
    pixmap.scaled(ui->online->size());
    ui->online->setPixmap(pixmap);
}

void chatwin::offline()
{
    //Êñ≠Á∫øÁä∂ÊÄÅ
    QPixmap pixmap("://res/icons/offline.png");
    pixmap.scaled(ui->online->size());
    ui->online->setPixmap(pixmap);
}
/*---------------------------------------------------------*/
void chatwin::dealMessage(ChatMessage *messageW, QListWidgetItem *item, QString text, QString time, QString id ,ChatMessage::User_Type type)
{
    ui->listWidget->addItem(item);
    messageW->setFixedWidth(ui->listWidget->width());
    QSize size = messageW->fontRect(text);
    item->setSizeHint(size);
    messageW->setText(text, time, size, id, type);
    ui->listWidget->setItemWidget(item, messageW);
}
void chatwin::dealMessageTime(QString curMsgTime)
{
    bool isShowTime = false;
    if(ui->listWidget->count() > 0) {
        QListWidgetItem* lastItem = ui->listWidget->item(ui->listWidget->count() - 1);
        ChatMessage* messageW = (ChatMessage *)ui->listWidget->itemWidget(lastItem);
        int lastTime = messageW->time().toInt();
        int curTime = curMsgTime.toInt();
        qDebug() << "curTime lastTime:" << curTime - lastTime;
        isShowTime = ((curTime - lastTime) > 60); // ‰∏§‰∏™Ê∂àÊÅØÁõ∏Â∑Æ‰∏ÄÂàÜÈíü
//        isShowTime = true;
    } else {
        isShowTime = true;
    }
    if(isShowTime) {
        ChatMessage* messageTime = new ChatMessage(ui->listWidget);
        QListWidgetItem* itemTime = new QListWidgetItem();
        ui->listWidget->addItem(itemTime);
        QSize size = QSize(ui->listWidget->width() , 40);
        messageTime->resize(size);
        itemTime->setSizeHint(size);
        messageTime->setText(curMsgTime, curMsgTime, size);
        ui->listWidget->setItemWidget(itemTime, messageTime);
    }
}

void chatwin::sendMsg(QString text)
{
    QString time = QString::number(QDateTime::currentDateTimeUtc().toTime_t());
    ChatMessage *message = new ChatMessage(ui->listWidget);
    QListWidgetItem *item = new QListWidgetItem();
    dealMessageTime(time);
    dealMessage(message, item, text, time, name ,ChatMessage::User_Me);
    ui->listWidget->verticalScrollBar()->setValue(ui->listWidget->verticalScrollBar()->maximum());

}

void chatwin::recvMsg(QString text)
{
    //ÊèêÁ§∫Èü≥
    QSound::play("://res/music/msg.wav");
    QString time = QString::number(QDateTime::currentDateTimeUtc().toTime_t());
    ChatMessage *message = new ChatMessage(ui->listWidget);
    QListWidgetItem *item = new QListWidgetItem();
    dealMessageTime(time);
    dealMessage(message, item, text, time, name ,ChatMessage::User_She);
    ui->listWidget->verticalScrollBar()->setValue(ui->listWidget->verticalScrollBar()->maximum());

}

/*--------------------ÊåâÈíÆ----------------------*/

void chatwin::on_close_btn_clicked()
{
    this->close();
}

void chatwin::on_min_btn_clicked()
{
    this->showMinimized();
}
/*-----------------Ê∂àÊÅØÊòæÁ§∫ÁïåÈù¢-------------------*/
void chatwin::on_enter_btn_clicked()
{
    QString msg = ui->plainTextEdit->toPlainText().trimmed();
    if(msg.size() == 0)
    {
        qDebug() << "empty";
        return;
    }
    qDebug()<<msg;
    ui->plainTextEdit->setPlainText("");
    emit resultReady_Me(msg);
    this->client_sent(msg);
}

/*--------------------Èü≥‰πêÊåâÈíÆ---------------------*/

void chatwin::on_mus_btn_clicked()
{
    QUrl url("http://goodtdgame.com/"); // ÊõøÊç¢‰∏∫‰Ω†ÊÉ≥Ë¶ÅÊâìÂºÄÁöÑÁΩëÈ°µÂú∞ÂùÄ
    QDesktopServices::openUrl(url);
}

/*-------------------Èº†Ê†áÊãñÂä®Á™óÂè£-------------------*/
void chatwin::mousePressEvent(QMouseEvent *event) {
    if (event->button() == Qt::LeftButton) {
        if (event->pos().y() < height() / 8) {  // Âè™Êúâ‰∏äÂçäÈÉ®ÂàÜÊâçÂÖÅËÆ∏ÊãñÂä®
            isMousePressed = true;
            dragPosition = event->globalPos() - frameGeometry().topLeft();
        }
    }
}

void chatwin::mouseMoveEvent(QMouseEvent *event) {
    if (isMousePressed) {
        if (event->pos().y() < height() / 8) {  // Âè™Êúâ‰∏äÂçäÈÉ®ÂàÜÊâçÂÖÅËÆ∏ÊãñÂä®
            emit dragSignal(event->globalPos() - dragPosition);
        }
    }
}

void chatwin::mouseReleaseEvent(QMouseEvent *event) {
    if (event->button() == Qt::LeftButton) {
        isMousePressed = false;
    }
}
/*--------------------------------------------------*/

